### PR TITLE
Allow buck http cache verify http header

### PR DESCRIPTION
**Pull request by zayhero**
Right now buck http cache didn't do any auth check against http header. To ensure we are safe from others uploading buggy artifacts, this PR is adding the logic to verify the header.

This PR includes:
1. config change to introduce `authorizationConfig`, so that user can send the `authorizedTokens` over
2. introduce `HttpHeaderAuthFilter ` to do http header authentication, what it does is extract the `Authorization` from Http header, and verify if that contains in `authorizedTokens` from config
3. change `BuckCacheApplication` to register the `HttpHeaderAuthFilter` so that it could be used
4. change `BuckCacheResource` to execute the authentication for **PUT** request
5. new unit test to verify the change

To use that, it require:
1. add authorized token into `http_write_headers` in `.buckconfig`, e.g.
```
[cache]
  mode = http
  http_mode = readwrite
  http_url = http://localhost:6457
  http_write_headers = Authorization: ZAYHERO
```
2. add the token (i.e. `ZAYHERO`) into `standalone.yml`, e.g.
```
authorizationConfig:
  authorizedTokens:
    - ZAYHERO
```

@fkorotkov

### CODE REVIEW

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    this.authorizedToken = authorizedToken;
+  }
+
+  @Override
+  public void filter(final ContainerRequestContext requestContext) throws IOException {
+    if (authorizedToken == null) {
+      return ;
+    }
+
+    String credential = requestContext.getHeaders().getFirst(HttpHeaders.AUTHORIZATION);
```
Comment (fkorotkov) [inline cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: should it be called token?

File: cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java [hunk 2]
```diff
@@ -111,7 +113,8 @@ public Response getCacheArtifact(@PathParam("key") String key) throws Exception
   @Path("key")
   @Consumes(MediaType.APPLICATION_OCTET_STREAM)
   public Response addArtifactToCache(PutCacheEntry putCacheEntry,
```
Comment (fkorotkov) [inline cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java]: is there a way to add a test for invoking `addArtifactToCache` without a token?

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java (hunk 1 above)
Comment (zayhero) [inline cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: sure :)

File: cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java (hunk 2 above)
Comment (zayhero) [inline cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java]: you mean not sending `AUTHORIZATION ` over AND test the `PUT` API? This part is already handled by `SingleServerIntegrationTest`, which doesn't contains any authorized token in its config file, and not sending any token via http header

Comment (fkorotkov) [inline cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java]: can you add a test that it rejects if a we are expecting a token? you have a unit test but an integration test will even better!

**Comment by zayhero** [comment]
@fkorotkov make these changes:
1. pass `authorizedTokens` into `SingleServerIntegrationTest`
2. update all existing tests to set token as header
3. add new test case `testCacheUploadFailAuthentication` to verify it will fail if missing token

**Comment by fkorotkov** [comment]
LGTM

**Pull request merged**